### PR TITLE
feat(nextjs,clerk-react,types,clerk-js): Add experimental support for `<Gate/>`

### PR DESCRIPTION
**Pull request by panteliselef**
## Description

This PR introduces a new public control component called `Gate` as   **experimental**.
- `<Experimental__Gate/>`  as client component
- `<Experimental__Gate/>`  as server component 
- Nexjts `auth()` function returns the `experimental__has` in order to check authorization server side
- `@clerk/backend`  returns the  `experimental__has` function in order to check authorization server side
- React `useAuth()` return the  `experimental__has` for checking authorization during SSR or CSR 

As of now the component will only support role based check, in the following weeks this will be updated to also support permission based checks which is the pattern we all our customers to use in the future.

**This PR will need to backported to v4 as it is complementary to "custom roles & permissions"**

## Checklist

- [ ] `npm test` runs as expected.
- [ ] `npm run build` runs as expected.
- [ ] (If applicable) [JSDoc comments](https://jsdoc.app/about-getting-started.html) have been added or updated for any package exports
- [ ] (If applicable) [Documentation](https://github.com/clerkinc/clerk-docs) has been updated

## Type of change

- [ ] 🐛 Bug fix
- [x] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

## Packages affected

- [ ] `@clerk/clerk-js`
- [ ] `@clerk/clerk-react`
- [ ] `@clerk/nextjs`
- [ ] `@clerk/remix`
- [ ] `@clerk/types`
- [ ] `@clerk/themes`
- [ ] `@clerk/localizations`
- [ ] `@clerk/clerk-expo`
- [ ] `@clerk/backend`
- [ ] `@clerk/clerk-sdk-node`
- [ ] `@clerk/shared`
- [ ] `@clerk/fastify`
- [ ] `@clerk/chrome-extension`
- [ ] `gatsby-plugin-clerk`
- [ ] `build/tooling/chore`

### CODE REVIEW

File: packages/nextjs/src/app-router/server/controlComponents.tsx [hunk 1]
```diff
@@ -13,3 +15,38 @@ export function SignedOut(props: React.PropsWithChildren) {
...
+  const handleFallback = () => {
+    if (!redirectTo && !fallback) {
+      throw new Error('Provide `<Gate />` with a `fallback` or `redirectTo`');
+    }
+
+    if (redirectTo) {
+      return redirect(redirectTo);
+    }
+
+    return <>fallback</>;
```
Comment (tmilewski) [inline packages/nextjs/src/app-router/server/controlComponents.tsx:44]: 🙃 Did you mean this?
```suggestion
    return <>{fallback}</>;
```

File: packages/nextjs/src/app-router/server/controlComponents.tsx [hunk 2]
```diff
@@ -13,3 +15,38 @@ export function SignedOut(props: React.PropsWithChildren) {
...
+  };
+
+  if (!isAuthorizedUser) {
+    return handleFallback();
+  }
+
+  return <>children</>;
```
Comment (tmilewski) [inline packages/nextjs/src/app-router/server/controlComponents.tsx:51]: 🙃 Similarly, did you mean?
```suggestion
  return <>{children}</>;
```

File: packages/types/src/jwtv2.ts [hunk 3]
```diff
@@ -17,6 +17,7 @@ export interface JwtHeader {
   alg: string;
   typ?: string;
   cty?: string;
+  // eslint-disable-next-line
```
Comment (tmilewski) [inline packages/types/src/jwtv2.ts:20]: ❓ Mind explicitly defining what we're disabling here? Thanks!

File: packages/react/src/hooks/utils.ts [hunk 4]
```diff
@@ -16,6 +16,7 @@ const clerkLoaded = (isomorphicClerk: IsomorphicClerk) => {
  * @internal
  */
 export const createGetToken = (isomorphicClerk: IsomorphicClerk) => {
+  // eslint-disable-next-line
```
Comment (tmilewski) [inline packages/react/src/hooks/utils.ts:19]: ❓ Mind explicitly defining what we're disabling here? Thanks!

File: packages/react/src/hooks/utils.ts [hunk 5]
```diff
@@ -29,6 +30,7 @@ export const createGetToken = (isomorphicClerk: IsomorphicClerk) => {
  * @internal
  */
 export const createSignOut = (isomorphicClerk: IsomorphicClerk) => {
+  // eslint-disable-next-line
```
Comment (tmilewski) [inline packages/react/src/hooks/utils.ts:33]: ❓ Mind explicitly defining what we're disabling here? Thanks!

**Review by tmilewski — CHANGES_REQUESTED** [review]
@panteliselef 👍 Nice! Added a few quick comments/questions.

File: packages/react/src/hooks/utils.ts (hunk 4 above)
Comment (panteliselef) [reply packages/react/src/hooks/utils.ts]: Before it was giving me an eslint error for using `explicit any`, now that's no longer the case (don't know why tho), so i'm removing this line completely

File: packages/react/src/hooks/utils.ts (hunk 5 above)
Comment (panteliselef) [reply packages/react/src/hooks/utils.ts]: same as above

File: packages/types/src/jwtv2.ts (hunk 3 above)
Comment (panteliselef) [reply packages/types/src/jwtv2.ts]: the line below no longer errors, so i'm removing this line

**Comment by panteliselef** [comment]
!snapshot

**Review by tmilewski — APPROVED** [review]

**Pull request merged**